### PR TITLE
Support remote for builtin operators in distributed autograd

### PR DESCRIPTION
**Pull request by mrshenli**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28656 Support remote for Python UDF in distributed autograd
* **#28630 Support remote for builtin operators in distributed autograd**

This includes:
1. Respect autograd context in rpc.remote for builtin ops
2. Force setting autograd context in RRef.to_here() even if the
message for to_here() does not contain any tensor.

Differential Revision: [D18138562](https://our.internmc.facebook.com/intern/diff/D18138562)

### CODE REVIEW

File: test/dist_autograd_test.py [hunk 1]
```diff
@@ -295,16 +302,70 @@ def _test_graph(self, fn):
...
+    @dist_init(setup_model_parallel=True)
     def test_graph_for_builtin_call(self):
         self._test_graph(torch.add)
 
-    @dist_init
+    @dist_init(setup_model_parallel=True)
     def test_graph_for_python_call(self):
         self._test_graph(my_py_add)
 
+    def _test_graph_remote(self, fn):
```
Comment (mrshenli) [inline test/dist_autograd_test.py]: Will consolidate this with the one for rpc_sync

File: test/dist_autograd_test.py [hunk 2]
```diff
@@ -125,7 +132,7 @@ def world_size(self):
     def init_method(self):
         return INIT_METHOD_TEMPLATE.format(file_name=self.file_name)
 
-    @dist_init
+    @dist_init(setup_model_parallel=True)
```
Comment (pritamdamania87) [inline test/dist_autograd_test.py]: https://github.com/pytorch/pytorch/pull/27940 will land soon, we shouldn't need to specify the default arg after that.

File: test/dist_autograd_test.py [hunk 3]
```diff
@@ -99,9 +100,15 @@ def _exec_func(self, exec_mode, method, *args):
...
-        else:
+        elif ExecMode.RPC_SYNC == exec_mode:
             return rpc.rpc_sync('worker{}'.format(self._next_rank()), method,
                                 args=(args))
+        elif ExecMode.REMOTE == exec_mode:
+            rref = rpc.remote('worker{}'.format(self._next_rank()), method,
+                              args=(args))
+            return rref.to_here().wait()
+        else:
+            raise ValueError("Unrecognized ExecMode {}".format(exec_mode))
```
Comment (pritamdamania87) [inline test/dist_autograd_test.py]: This is awesome! Thanks for refactoring dist_autograd_test in such a clean way to enable RRef tests.

**Comment by zhaojuanmao** [comment]
overall lgtm

but after this diff, making me think whether we should send autograd meta, create autograd context and check whether we should attach autograd function for every rpc call in default. that will simplify our codes a lot.

File: test/dist_autograd_test.py [hunk 4]
```diff
@@ -536,20 +576,21 @@ def _verify_backwards_remote(self, tensors, context_id, local_grads, *args):
...
         local_grads = None
         t1 = torch.rand((3, 3), requires_grad=True)
         t2 = torch.rand((3, 3), requires_grad=True)
-        for exec_mode in [ExecMode.LOCAL, ExecMode.REMOTE]:
+        for exec_mode in [ExecMode.LOCAL, ExecMode.RPC_SYNC, ExecMode.REMOTE]:
             with dist_autograd.context() as context_id:
                 ret = self._exec_func(exec_mode, torch.add, t1, t2)
                 loss = ret.sum()
-                local_grads = self._verify_backwards(exec_mode, [loss], context_id, local_grads, t1, t2)
+                ret = self._verify_backwards(exec_mode, [loss], context_id, local_grads, t1, t2)
```
Comment (pritamdamania87) [inline test/dist_autograd_test.py]: Shouldn't we also verify here (and in other tests) that a gradient is accumulated on the OwnerRRef?

File: test/dist_autograd_test.py [hunk 5]
```diff
@@ -735,12 +780,12 @@ def test_backward_invalid_args(self):
                 t = torch.rand(1, requires_grad=True)
                 dist_autograd.backward([t])
 
-    @dist_init
+    @dist_init(setup_model_parallel=True)
```
Comment (pritamdamania87) [inline test/dist_autograd_test.py]: Can we add the following tests:

1) Never call RRef.to_here() and verify grads are accumulated on all tensors except the RRef tensor.
2) A creates an RRef on B and shares that RRef with C. C now calls to_here() and does some computation and returns the result to A. A now calls backward on the result. We should verify the OwnerRRef on B accumulates gradients appropriately.
3) Multiple nodes call RRef.to_here() on the same rref and perform some computation. The backward pass should result in multiple edges to the same RRef and appropriate gradient accumulation. We could denote one node as the parameter server for this test and retrieve the RRef to do some computation.

File: torch/csrc/distributed/autograd/utils.cpp [hunk 6]
```diff
@@ -42,37 +42,45 @@ DistAutogradContext* addRecvRpcBackward(
...
   auto& autogradContainer = DistAutogradContainer::getInstance();
   DistAutogradContext& autogradContext =
       autogradContainer.getOrCreateContext(autogradMetadata.autogradContextId);
 
-  // Attach the tensors as inputs to the autograd function.
-  auto grad_fn = std::make_shared<RecvRpcBackward>(
-      autogradMetadata, autogradContext, fromWorkerId);
-  for (auto& tensor : tensors) {
-    torch::autograd::set_history(tensor, grad_fn);
+  if (!tensors.empty()) {
```
Comment (pritamdamania87) [inline torch/csrc/distributed/autograd/utils.cpp]: Did we catch this issue with the unit tests you added? If not can we add a unit test?

File: torch/csrc/distributed/autograd/utils.h [hunk 7]
```diff
@@ -34,19 +34,23 @@ TORCH_API DistAutogradContext* addRecvRpcBackward(
...
+// context and tensors require grads or checkRequiresGrad is false, in this
+// case, return RpcWithAutograd message; otherwise return original rpc message.
+// NB: checkRequiresGrad is useful when the request does not contain any tensor
+// but the corresponding response does.
 TORCH_API rpc::Message getMessageWithAutograd(
     const rpc::worker_id_t dstId,
     rpc::Message&& wrappedRpcMsg,
-    rpc::MessageType msgType);
+    rpc::MessageType msgType,
+    bool checkRequiresGrad = true);
```
Comment (pritamdamania87) [inline torch/csrc/distributed/autograd/utils.h]: nit: maybe call this `forceGradRecording` and default it false. This way `ScriptRRefFetchCall` can set this to true, to force recording grad even if tensors are empty.

File: torch/csrc/distributed/rpc/rref.cpp [hunk 8]
```diff
@@ -133,14 +136,36 @@ template <>
...
-      ScriptRRefFetchCall(rrefId()).toMessage());
+      ScriptRRefFetchCall(ownerId_, rrefId()).toMessage(),
+      false /* checkRequiresGrad */);
 
   futureResponse->addCallback([future](const Message& message) {
     RRefContext::handleException(message);
-    auto rfr = ScriptRRefFetchRet::fromMessage(message);
-    future->markCompleted(rfr->values().front());
+    if (message.type() == MessageType::FORWARD_AUTOGRAD_RESP) {
+      auto response = deserializeResponse(message);
```
Comment (pritamdamania87) [inline torch/csrc/distributed/rpc/rref.cpp]: nit: This deserializeResponse can be outside the if statement right? Then the result can be used in either of the if-else clauses?

**Review by pritamdamania87 — CHANGES_REQUESTED** [review]
Looks great! Mostly requesting changes for a few additional unit tests.

File: test/dist_autograd_test.py (hunk 2 above)
Comment (mrshenli) [reply test/dist_autograd_test.py]: Yes, I will rebase after that lands.

File: torch/csrc/distributed/autograd/utils.cpp (hunk 6 above)
Comment (mrshenli) [reply torch/csrc/distributed/autograd/utils.cpp]: Yes, the `to_here()` message is wrapped with autograd, which will trigger this method on the caller. However, we should not insert a recv autograd function for it because that message does not contain any tensor.

File: test/dist_autograd_test.py (hunk 4 above)
Comment (mrshenli) [reply test/dist_autograd_test.py]: The result tensor of `torch.add` (owner) actually does not require grad, but let me add another test to cover that.

**Review by pritamdamania87 — APPROVED** [review]
